addons/survey/tests/test_survey.py

```diff
@@ -0,0 +1,227 @@
+
+    def test_04_question_numerical_box(self):
+        question = self.Question.sudo(self.survey_manager).create({
+            'page_id': self.page1.id, 'question': 'Q0', 'type': 'numerical_box', 'validation_required': True,
+            'validation_min_float_value': 2.1, 'validation_max_float_value': 3.0, 'validation_error_msg': "Error"})
+        answer_tag = '%s_%s_%s' % (self.survey1.id, self.page1.id, question.id)
+        results = [('aaa', _('This is not a number')), ('4.5', 'Error'), ('0.1', 'Error')]
+        for i in range(len(results)):
+            self.assertEqual(self.Question.validate_question(question, {answer_tag: results[i][0]}, answer_tag), {answer_tag: results[i][1]}, msg="\
+                Validation function for type numerical_box is unable to notify if answer is vaiolating the validation rules")
+
+    def test_05_question_datetime(self):
+        question = self.Question.sudo(self.survey_manager).create({
+            'page_id': self.page1.id, 'question': 'Q0', 'type': 'datetime', 'validation_required': True,
+            'validation_min_date': '2015-03-20 00:00:00', 'validation_max_date': '2015-03-25 00:00:00', 'validation_error_msg': "Error"})
+        answer_tag = '%s_%s_%s' % (self.survey1.id, self.page1.id, question.id)
+        results = [('2015-55-10', _('This is not a date/time')), ('2015-03-19 00:00:00', 'Error'), ('2015-03-26 00:00:00', 'Error')]
+        for i in range(len(results)):
+            self.assertEqual(self.Question.validate_question(question, {answer_tag: results[i][0]}, answer_tag), {answer_tag: results[i][1]}, msg="\
+                Validation function for type datetime is unable to notify if answer is vaiolating the validation rules")
+
+    def test_06_survey_sharing(self):
+        surveys = [
+            self.Survey.sudo(self.survey_manager).create({  # Survey without any page or question.
+                'title': "Test survey"}),
+            self.Survey.sudo(self.survey_manager).create({  # Closed Survey.
+                'title': "S0", 'stage_id': self.Stage.search([('closed', '=', True)]).id,  # Getting Closed stage id.
+                'page_ids': [(0, 0, {'title': "P0", 'question_ids': [(0, 0, {'question': "Q0", 'type': 'free_text'})]})]})]
+        for survey in surveys:
+            self.assertRaises(UserError, survey.action_send_survey)
+
+    def test_07_survey_email_message(self):
+        mail_messages = [
+            self.SurveyMailMessage.sudo(self.survey_manager).create({  # Mail Message without __URL__ in body.
+                'survey_id': self.survey1.id, 'public': 'email_public_link'}),
+            self.SurveyMailMessage.sudo(self.survey_manager).create({  # Mail Message without recipents.
+                'survey_id': self.survey1.id, 'public': 'email_public_link', 'body': "__URL__"})]
+        for message in mail_messages:
+            self.assertRaises(UserError)
```

Comment:
it would be nice to add test for all action button which should return action with valid URL
current coverage status for addons/survey/survey : 51%